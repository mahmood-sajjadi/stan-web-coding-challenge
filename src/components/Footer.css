#footer-nav li + li::before {
  content: '|';
  padding: 0 5px;
}

#footer-nav a:visited {
  color: inherit;
}

#footer-stores li + li::before{
  content: '';
  padding: 0 5px;
}

#footer-social li + li::before {
  content: '';
  padding: 0 5px;
}

#footer-social li img:hover {
  opacity: 0;
  transition: opacity 0.3s linear;
}