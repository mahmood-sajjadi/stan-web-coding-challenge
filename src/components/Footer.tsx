import { useMemo } from "react";
import './Footer.css'
import { Link } from "react-router-dom";
import type { FooterMenuItemType, SocialItemType, StoreItemTypes } from "../config";

type Props = {
    footerStyle?: React.CSSProperties;
    menuItems: FooterMenuItemType[];
    copyright: string;
    socials: SocialItemType[];
    stores: StoreItemTypes[]
}

const footerStyleBase = {
    backgroundColor: 'rgb(30, 30, 30)',
    display: 'block',
    width: 'calc(100% - 200px)',
    color: 'rgb(195, 192, 187)',
    fontSize: '10px'
}

const menuStyle = {
    display: 'flex',
    alignItems: 'center',
    padding: '0',
}

const menuItemStyle = {
    listStyleType: 'none',
    display: 'contents',
    cursor: 'pointer',
}

const menuItemLinkStyle = {
    textDecoration: 'none',
}

const iconStyle = {
    height: '100%',
}

const socialMediaStyle = {
    ...menuStyle,
    float: 'left',
}

const socialMediaItemStyle = {
    listStyleType: 'none',
    display: 'block',
    cursor: 'pointer',
    height: '30px',
    backgroundPosition: 'right center',
    backgroundRepeat: 'no-repeat',
    backgroundSize: 'contain',
}

const storeStyle = {
    ...menuStyle,
    float: 'right',
}

const storeItemStyle = {
    listStyleType: 'none',
    display: 'block',
    cursor: 'pointer',
    height: '30px',
}

const socialAndStoreRowStyle = {
    marginTop: '50px',
}

function menuItemWithLink(title: string, url: string) {
    return (
        <Link to={url} style={menuItemLinkStyle}>
            {title}
        </Link>
    )
}

function menuItemsList(menuItems: FooterMenuItemType[]) {
    return (
        <nav>
        <ul id="footer-nav" style={menuStyle}>
            {
                menuItems.map(item => (
                    <li key={item.title} style={menuItemStyle}>
                        {item.url ? menuItemWithLink(item.title, item.url) : item.title}
                    </li>
                ))
            }
        </ul>
    </nav>
    )
}

function socialIconList(socials: SocialItemType[]) {
    return (
        <ul id="footer-social" style={socialMediaStyle}>
        {
            socials.map(item => {
                const style = {
                    ...socialMediaItemStyle,
                    backgroundImage: `url(${item.images.hover})`,
                }
                return (
                    <li key={item.title} style={style}>
                        <img src={item.images.normal} alt={item.title} style={iconStyle} />
                    </li>
                )
            })
        }
    </ul>
    )
}

function storeIconsList(stores: StoreItemTypes[]) {
    return (
        <ul id="footer-stores" style={storeStyle}>
        {
            stores.map(item => (
                <li key={item.title} style={storeItemStyle}>
                    <img src={item.image} alt={item.title} style={iconStyle} />
                </li>
            ))
        }
        </ul>
    );
}
export function Footer(props: Props) {
    const { footerStyle, menuItems, copyright, socials, stores } = props;

    const footerWrapperStyle = useMemo(() => ({
        ...footerStyleBase,
        ...footerStyle,
    }), [footerStyle]);

    const menu = useMemo(() => menuItemsList(menuItems), [menuItems]);

    const socialIcons = useMemo(() => socialIconList(socials), [socials]);

    const storeIcons = useMemo(() => storeIconsList(stores), [stores]);

    return (
        <footer style={footerWrapperStyle}>
            {menu}
            {copyright}
            <div style={socialAndStoreRowStyle}>
                {socialIcons}
                {storeIcons}
            </div>
        </footer>
    );
}