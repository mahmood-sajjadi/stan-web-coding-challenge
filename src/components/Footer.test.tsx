import { render } from '@testing-library/react'
import {Footer} from './Footer'

describe('Footer', () => {
  it('renders the Footer component', () => {
    const footer = render(
        <Footer
            copyright='copyright'
            menuItems={[{title: 'test'}]}  
            socials={[{images: {normal: '', hover: ''}, title: 'test'}]}
            stores={[{title: 'test', image: ''}]}
        />
    );
    expect(footer).toMatchSnapshot();
  })
})