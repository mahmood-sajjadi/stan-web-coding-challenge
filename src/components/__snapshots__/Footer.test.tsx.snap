// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Footer > renders the Footer component 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <footer
        style="background-color: rgb(30, 30, 30); display: block; width: calc(100% - 200px); color: rgb(195, 192, 187); font-size: 10px;"
      >
        <nav>
          <ul
            id="footer-nav"
            style="display: flex; align-items: center; padding: 0px;"
          >
            <li
              style="list-style-type: none; display: contents; cursor: pointer;"
            >
              test
            </li>
          </ul>
        </nav>
        copyright
        <div
          style="margin-top: 50px;"
        >
          <ul
            id="footer-social"
            style="display: flex; align-items: center; padding: 0px; float: left;"
          >
            <li
              style="list-style-type: none; display: block; cursor: pointer; height: 30px; background-position: right center; background-repeat: no-repeat; background-size: contain; background-image: url("");"
            >
              <img
                alt="test"
                style="height: 100%;"
              />
            </li>
          </ul>
          <ul
            id="footer-stores"
            style="display: flex; align-items: center; padding: 0px; float: right;"
          >
            <li
              style="list-style-type: none; display: block; cursor: pointer; height: 30px;"
            >
              <img
                alt="test"
                style="height: 100%;"
              />
            </li>
          </ul>
        </div>
      </footer>
    </div>
  </body>,
  "container": <div>
    <footer
      style="background-color: rgb(30, 30, 30); display: block; width: calc(100% - 200px); color: rgb(195, 192, 187); font-size: 10px;"
    >
      <nav>
        <ul
          id="footer-nav"
          style="display: flex; align-items: center; padding: 0px;"
        >
          <li
            style="list-style-type: none; display: contents; cursor: pointer;"
          >
            test
          </li>
        </ul>
      </nav>
      copyright
      <div
        style="margin-top: 50px;"
      >
        <ul
          id="footer-social"
          style="display: flex; align-items: center; padding: 0px; float: left;"
        >
          <li
            style="list-style-type: none; display: block; cursor: pointer; height: 30px; background-position: right center; background-repeat: no-repeat; background-size: contain; background-image: url("");"
          >
            <img
              alt="test"
              style="height: 100%;"
            />
          </li>
        </ul>
        <ul
          id="footer-stores"
          style="display: flex; align-items: center; padding: 0px; float: right;"
        >
          <li
            style="list-style-type: none; display: block; cursor: pointer; height: 30px;"
          >
            <img
              alt="test"
              style="height: 100%;"
            />
          </li>
        </ul>
      </div>
    </footer>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
